### ISSUE TITLE
Run a single node test on each node in a partition

### ISSUE THREAD

**Issue by casparvl** [eth-cscs/reframe#2395]
I'd like to run a single node test on each node in a partition, but couldn't find an elegant way to do this. A short discussion on the ReFrame General slack channel led to an implementation that does:

```
node = parameter(['node1', 'node2'])

@run_before('run')
    def prepare_run(self):
        self.job.options = [f"--nodelist={self.node}"]
        self.name = f "{self.name}-{self.node}"
```

but it requires hardcoding the nodenames, or at the very least keeping the nodenames (per partition) somewhere in a list.

One option would be to add a property (e.g. 'nodes' or 'nodenames') to partition in ReFrame's settings file that holds all the nodenames. That way, it is still hardcoded, but at least it is in one central place (no duplication between tests). Plus, with a utility function that does the reverse of https://reframe-hpc.readthedocs.io/en/stable/utility_functions_reference.html?highlight=nodelist#reframe.utility.nodelist_abbrev , one would only have to specify the relevant range(s) in the `nodenames`. E.g.

```
'systems': [
...
   'partitions': [
...
       'nodenames': ['nid[100-200]', 'nid[400-500]']
...
```

Starting with support for something like this in the settings file, functionality could then later easily be extended by a dedicated feature that essentially does the above (i.e. create a parameter based on the nodenames, and pass additional job options), but bases the `node` parameter on the content of `nodenames` for the current partition (and with proper support for various job backends instead of hard-coded arguments).

Instead of coding a nodelist in the settings file, we could automatically try to query it from the job scheduler. However, I don't think this approach would match with ReFrame's concept of partitions: partitions in the ReFrame settings file don't always match 1-to-1 with partitions in SLURM (and this actually gives a lot of flexibility). Thus, I'd personally prefer the approach with hardcoded nodelists in the ReFrame settings file, as long as it supports defining node ranges.

**Comment by vkarak** [comment]
@casparvl I don't think that specifying the node list in the partition configuration is a good idea, because it can get outdated and very often you may have nodes offline, so you can't simply parametrize on such a static set of nodes. The job scheduler internal interface gives you the possibility to select the nodes of a specific _ReFrame_ partition. The Slurm backend in particular is smart enough to interpret any job option (not just the `-p partition`) so that it can actually select the nodes of a ReFrame partition based on the `access` options plus any other options (constraints, nodelists, reservations etc.) are passed from the command line.

The problem here is more fundamental and I think it is the same as in #2334. Even if you get the list of nodes of interest in a partition, you would still need to parametrize your test differently per partition. This is not currently possible at the definition of a parameter, but the workaround is to define the parameter to take the union of the parameter values and then filter in a post-init hook. For nodes, I agree, this is not so nice.

As soon as #2334 is fixed, you will be able to select any set of tests from the command line and launch them on all or part of nodes of a reframe partition.

I will mark this issue as  a duplicate, but if you think it's not, we can further discuss it. I will write some more details in #2334.
